### DISCUSSION TITLE
Is Android client dev still active?

### DISCUSSION THREAD

**Discussion by Andrew999999** [jamulussoftware/jamulus#1142] [General]
A while ago there was some discussion about an Android client, and it appeared that it was close to completion.  Since then I've not read anything.  Is this still an active project area?  I'd be very interested to know, as it has the potential to significantly expand the usefulness of Jamulus by allowing much more inclusive access to sessions.  Thanks.

**Comment by ann0see** [comment]
The android client works but with high latency. You can even download the .apk files once the next release is out or: https://github.com/jamulussoftware/jamulus/releases/tag/r3_7_0beta3

**Reply by Andrew999999** [reply]
That's great - thanks - I'll do some testing.

**Comment by dcorson-ticino-com** [comment]
There may be many Android devices out there, but as far as I know they do not have wired Internet access.
While using wifi with Jamulus may work, it is far from being good.
I worry that many people using devices that can only connect to the Internet using wifi will be bad for Jamulus' reputation as I am sure that most people will blame the poor audio quality on the Jamulus program and not on the fact that the devices are inappropriately connected to the Internet.
(I note that in my surroundings ALL the people who started using Jamulus with wifi claiming they had the best wifi in the world have gone to using a wired connection.  The difference in sound quality is not a myth, but is substantial.)

**Reply by Andrew999999** [reply]
I agree that wired is much better - but I've achieved acceptable results with some people who have been forced to use wi-fi.   I certainly think it's worth trying.

**Reply by ann0see** [reply]
I tested LTE on an Android phone and it seems to work much better then with WIFI.

**Comment by gene96817** [comment]
I would make the distinction between an Android smartphone and a tablet.  I see there are Ethernet adapters sold for Andriod tablets. I hope any work on an Android client is using an ethernet adapter.

**Reply by Andrew999999** [reply]
Good point.  Some smartphones will also support OTG to Ethernet adapters via their "charging" USB socket, depending on the manufacturer.

**Comment by gene96817** [comment]
@Andrew999999 

> but I've achieved acceptable results

It would be good to know what you consider acceptable results. Best if you report how much of the overall delay is from the Android client. Remember this delay is added to all other sources so what is acceptable is dependent on the other factors.

**Reply by Andrew999999** [reply]
Using Jamulus server on the same gigabit LAN I get minimal ping time (2-3mS) and overall delay via wi-fi of around 28-50mS from the Android client.  However the overall delay varies enormously from one moment to another.  The lower end of the range comes from a smartphone using an octa-core 2.8GHz processor.  The higher end from a tablet using an octa-core 1.9GHz processor.  Both were using Android 10.  Wi-fi connection speed for both was about 250 Mbps.  I'm using the Release 3.6.2dev-6e630f4 (r3_7_0beta3) Android client.

For my typical remote private servers I see about 15 mS ping time, bringing overall delay up to 45-65 mS.  From subjective testing this is _just about_ acceptable to play along with others and not loose speed - but a little high for a truly pleasant experience.  For comparison, using Windows 10 and wired gigabit ethernet with the same private servers, I see overall delay of about 32 mS which is subjectively absolutely fine.

I used the same high quality headphones with all devices - and the audio quality via Android was just as fantastic as the Windows device.  I used "High" audio quality "Stereo" settings for all.  The reported quality of the internal microphones was "satisfactory" with a little lack of bass at distance - presumably because they're tuned for the voice being close to the mic.  With an instrument (I used violin for these tests) the quality was reported as perfectly acceptable.

So, in summary, via Wi-Fi, a phone or tablet was acceptable via wi-fi for short periods and gave excellent audio quality with slightly annoying delay.

I have an OTG Ethernet adapter arriving today so I'll test the difference with that.

NOTE - I also tried attaching the headphones via Bluetooth 5 to the tablet/phone.  Not surprisingly this totally killed the experience with almost a second overall delay.  And of course this additional device to headphones delay is not reported by Jamulus. This is a consideration because, increasingly, devices use a single USB-C port for power/headphones - and potentially OTG Ethernet.

**Reply by ann0see** [reply]
Thank you for this response (I kind of expected this outcome). So probably the main issue is WiFi and to some extent latency (is it related to power saving options or just because Oboe is slow?). Running Jamulus on a Raspberry Pi 0 does work and give acceptable results.

I think, Jamulus over 5G or 4G might work better (at least that's my experience with an android device tested a while ago).

**Reply by Andrew999999** [reply]
So I tested the same setup with an OTG gigabit ethernet adapter plugged in to the USB-C port of the Android devices.  Not only is the overall delay on Android 10 much more comparable to what i see from Windows 10, but the ping and delay times are much more stable.  The tablet or phone both become very usable.

Of course this is still a limited use case - it is not possible to attached an external audio interface, plug in an external microphone or instrument, or use sampled instruments e.g. via midi.  But it would suite those using acoustic instruments or voice.

What I did find is that for all devices the microphone is effective at a distance - so the device could be placed on e.g. a music stand and still the instrument/voice balances well.  As mentioned before, moving away from the device microphone in this way does noticeably reduce bass response, however.

**Reply by ann0see** [reply]
> it is not possible to attached an external audio interface

Are you sure about that? What happens if you connect an USB Hub with LAN and Interface?

**Reply by Andrew999999** [reply]
I haven't tested that. Are the drivers built in to Android?

**Reply by ann0see** [reply]
Not sure but it’s worth a try. I know that iOS supports many interfaces natively.

**Reply by gene96817** [reply]
I know the following comment steps further away from simplicity....
What happens if we use a USB hub for the ethernet interface and the microphone and, maybe, even the headset? It will look like a monstrosity but that is the natural consequence of shrinking a notebook to tablet to a smartphone. 

If it mostly works (or can work) I suggest this could be our reference development configuration for all the external hardware.

**Reply by gene96817** [reply]
> Are you sure about that? What happens if you connect an USB Hub with LAN and Interface?

> Are the drivers built into Android?

One of the support challenges will be that the hardware manufacturers build custom packages for their hardware models. They could be omitting components that are useless to their product manager's goals.

**Comment by ann0see** [comment]
@nefarius2001 has some more experience than I have (he made the automatic Android build possible).

**Reply by Andrew999999** [reply]
Thanks. Testing going well so far. However I notice that I can't add the server port in the usual way (by adding ':portnumber' after the server name. This means I can't connect to any of my private servers, none of which use the default port number.

**Reply by ann0see** [reply]
Strange. Can you open an issue on that?

**Reply by Andrew999999** [reply]
I'll do some more testing first.

**Comment by gene96817** [comment]
I am uncomfortable speaking about acceptable performance using wireless because there are so many factors that most people don't think about. Whether it is wifi, Bluetooth, or 4G/5G, the geometry (location of the devices) and the environment (especially devices not under the user's ownership and control) affect performance. If you understand how wireless works, the environmental issues are dominated by the media access protocols (i.e. when they can transmit) that all the nearby devices share. The performance variances are normal and should be expected. The challenge is how do we include these things as acceptable performance.

**Reply by Andrew999999** [reply]
I agree - perhaps a better description for Jamulus via wi-fi would be "unpredictable" performance, and certainly not recommended.

**Comment by gene96817** [comment]
Can we report the performance for the Andriod release using an ethernet adapter? The two major benefits would be, we establish the best result from the Android release and we remind the Android users, yes, you can use ethernet. We are likely to see a lot of performance variations depending on the manufacturer and model. I personally don't want to be researching why someone with an inexpensive Andriod phone has "mysterious" performance issues. It's awesome you bought a brand new $50 Android phone, but...

**Comment by herbetom** [comment]
I just did a quick test of the latest Preview Release on my OnePlus 6t (Android 10) (https://github.com/jamulussoftware/jamulus/releases/download/r3_7_0beta3/jamulus_3.6.2dev-6e630f4_android.apk)

I found that while the received Audio from my Laptop seems to be good on the Android Client and the local playback is also about right, the received audio on my Laptop from the Android Client is really distorted.

Additionally with activated gesture navigation the App seems to crash sometimes (or at least it get's closed and the connection is lost, other then just a quick trip to the homescreen and back) if i use the return gesture to get out of the Chat window.

**Reply by Andrew999999** [reply]
I get the same issue with gesture navigation.  My phone doesn't have traditional buttons, and the gesture for "return" is not recognised.  Also I can not see a "close" or "cancel" button on the Chat window, so cannot return from it.  I'll raise this as an issue.

**Comment by sthenos** [comment]
I regularly use the android build on my samsung S8 with an ethernet adapter or wifi, depending on the mood i'm on.  I get pretty good latency to my london server.  Around 29ms overall delay or so and I am able to jam and perform with musicians regularly using it.     I can recommend it if anyone wants to use it

**Reply by gene96817** [reply]
I have a few users that would like to try it. How do I get the software? What is the install procedure?

**Reply by ann0see** [reply]
Have a look at the website Repo in the wiki/en/ Installation-for-Android file. It's different for every device but quite easy if you have an apk file

**Reply by Andrew999999** [reply]
@gene96817 I'd check that the device each user has actually supports an OTG Ethernet device.  Most tablets will, but some phones only support charging on their USB ports.  As I reported earlier, overall delay without a plugged in ethernet cable significantly increases overall delay and makes Jamulus quite difficult to use - this isn't recommended.

@sthenos presumably you're located in or near to London.  I'm located in Scotland and overall delay via wifi to a London server is too high with Android to be useable.

**Reply by gene96817** [reply]
@Andrew999999   I certainly don't like the idea of having a class of users (e.g. Android) that only use wifi. This will signal to many users that ethernet is optional. 

I have had some distant visitors to our rehearsal. The extra delay is distracting, although we do enjoy the fellowship with distant musicians. I find it a bit rude that some of them chose to visit with a wi-fi connection. I don't want to be tolerate the extra latency because of their choice of convenience. It would be aweful to have all the Android users with the extra latency.
